app/src/main/java/com/codesoom/assignment/product/service/ProductService.java

```diff
@@ -0,0 +1,42 @@
+package com.codesoom.assignment.product.service;
+
+import com.codesoom.assignment.product.domain.Product;
+
+import java.util.List;
+
+public interface ProductService {
+    /**
+     * 제품 목록 전체를 반환한다.
+     * @return 제품 목록 리스트
+     */
+    List<Product> getProducts();
+
+    /**
+     * 식별자와 일치하는 제품을 반환한다
+     * @param id 제품 식별자
+     * @return 제품
+     */
+    Product findProductById(Long id);
```

Comment:
혹시 인터페이스에서 선언하지 않은 예외를 구현체에서 멋대로 던지고 있지는 않은가요?
```suggestion
    /**
     * 식별자와 일치하는 제품을 반환한다
     * @param id 제품 식별자
     * @return 제품
     */
    Optional<Product> findProductById(Long id);
```

Reply:
앗 그렇네요

이 부분에서 반환 값에 전부(Service, Controller) ```Optional```을 씌워주라는 말씀이신가요 아니면 인터페이스에 예외 선언을 하라는 말씀이신가요??

만약 인터페이스에서 
``` java
Product findProductById(Long id) throws ProductNotFoundException;
```
이런식으로 하면 Optional로 감싼 객체가 아닌 그냥 Product를 반환해도 되는걸까요?

Reply:
이 메소드 이름을 두고 잘 생각해 봅시다. 이건 아이디를 통해 상품을 검색하는 메소드라 생각할 수 있어요. 웹 브라우저에서 뭔가 검색했을 때 검색 결과가 안 나오면 보통 어떻게 되나요? 예외가 발생하면서 웹 브라우저가 종료될까요? 검색 결과가 없다면 없는 상태로 리턴해주는 것도 괜찮은 방법입니다. 예외는 조심스럽게, 예외적으로 사용하는 것이 바람직합니다. 예외를 제어 수단으로 남용하는 것이 좋지 않다는 것을 염두에 두고 항상 예외를 최소화하면서 코딩하는 습관을 길러보세요. (물론 예외보다 null을 리턴하는 것이 일반적으로 더 안 좋은 방법입니다.)

Reply:
이해 했습니다!